x-pack/plugins/cases/server/connectors/cases/retry_service.test.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { CasesConnectorError } from './cases_connector_error';
+import { CaseConnectorRetryService } from './retry_service';
+import type { BackoffFactory } from './types';
+
+describe('CryptoService', () => {
+  const nextBackOff = jest.fn();
+  const cb = jest.fn();
+
+  const backOffFactory: BackoffFactory = {
+    create: () => ({ nextBackOff }),
+  };
+
+  let service: CaseConnectorRetryService;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    nextBackOff.mockReturnValue(1);
+    service = new CaseConnectorRetryService(backOffFactory);
+  });
+
+  it('should not retry if the error is not CasesConnectorError', async () => {
+    cb.mockRejectedValue(new Error('My error'));
+
+    await expect(() => service.retryWithBackoff(cb)).rejects.toThrowErrorMatchingInlineSnapshot(
+      `"My error"`
+    );
+
+    expect(cb).toBeCalledTimes(1);
+    expect(nextBackOff).not.toBeCalled();
+  });
+
+  it('should not retry if the status code is not supported', async () => {
+    cb.mockRejectedValue(new CasesConnectorError('My case connector error', 500));
+
+    await expect(() => service.retryWithBackoff(cb)).rejects.toThrowErrorMatchingInlineSnapshot(
+      `"My case connector error"`
+    );
+
+    expect(cb).toBeCalledTimes(1);
+    expect(nextBackOff).not.toBeCalled();
+  });
+
+  it('should not retry after trying more than the max attempts', async () => {
+    const maxAttempts = 3;
+    service = new CaseConnectorRetryService(backOffFactory, maxAttempts);
+
+    cb.mockRejectedValue(new CasesConnectorError('My transient error', 409));
+
+    await expect(() => service.retryWithBackoff(cb)).rejects.toThrowErrorMatchingInlineSnapshot(
+      `"My transient error"`
+    );
+
+    expect(cb).toBeCalledTimes(maxAttempts + 1);
```

Comment:
Why `maxAttempts + 1`?

Reply:
The execution goes as:

1. The `cb` is called for the first time. `cb` throws an error. The retry service retries the `cb`.
2. The `cb` is called for a second time. First rety. `cb` throws an error. The retry service retries the `cb`.
3. The `cb` is called for a third time. Second retry. `cb` throws an error. The retry service retries the `cb`.
4. The `cb` is called for the fourth time. Third retry. `cb` throws an error. The retry service does not retry and throws an error.

Basically the first execution of `cb` does not count as a retry.